src/metabase/lib/equality.cljc

```diff
@@ -104,106 +109,365 @@
+  ([needles haystack-metadatas]
+   (closest-matches-in-metadata needles haystack-metadatas {}))
+
+  ([needles haystack-metadatas opts]
+   (when (seq haystack-metadatas)
+     ;; The matching is done in two passes: first as plain refs, second by forcing ID refs.
+     ;; Each of these returns a map of `{^{::index i} haystack-ref needle-index}`. Those haystack values which match
+     ;; are removed between stages.
+     (cascading-matches
+       needles haystack-metadatas opts
+       [;; The first pass attempts to match the refs using the regular [[find-closest-matches-for-refs]].
+        lib.ref/ref
+        ;; If that fails to match some needles, and there's at least some needles which are "Field ID" refs like
+        ;;
+        ;;    [:field {} 1]
+        ;;
+        ;; then try again, forcing creation of Field ID refs for all of the metadatas. This way we can match
+        ;; things where the FE incorrectly used a Field ID ref even tho we were expecting a nominal :field
+        ;; literal ref like
+        ;;
+        ;;    [:field {} "my_field"]
+        (when (some field-id-ref? needles)
+          ;; Force creation of a Field ID ref by giving it a `:lib/source` that will make the ref creation code
+          ;; treat it as coming from a source table. This only makes sense for metadatas that have an associated
+          ;; Field `:id`, so others are left as-is.
+          (fn [metadata]
+            (cond-> metadata
+              (:id metadata) (assoc :lib/source :source/table-defaults)
+              metadata       lib.ref/ref)))])))
+
+  ([query stage-number needles haystack-metadatas]
+   (closest-matches-in-metadata query stage-number needles haystack-metadatas {}))
+
+  ([query                 :- [:maybe ::lib.schema/query]
+    stage-number          :- :int
+    needles               :- [:sequential [:maybe ::lib.schema.ref/ref]]
+    haystack-metadatas    :- [:sequential lib.metadata/ColumnMetadata]
+    opts                  :- [:map [:keep-join? {:optional true} :boolean]]]
+   ;; This could be more efficient if the second pass were run on the un-matched subset, like cascading-matches.
+   (let [basic   (closest-matches-in-metadata needles haystack-metadatas opts)
+         ;; In case that fails, we need to fall back and try matching any integer IDs by name like the four-arity
+         ;; version of [[find-closest-matching-ref]].
+         by-name (when (and query
+                            (some field-id-ref? needles))
+                   (closest-matches-in-metadata
+                     (for [[_field _opts field-id] needles]
+                       (when (pos-int? field-id)
+                         (lib.ref/ref (resolve-field-id query stage-number field-id))))
+                     haystack-metadatas
+                     opts))]
+     ;; Prefer the more precise integer ID matches over the name-based ones.
+     (merge by-name basic))))
```

Comment:
Would you pass in `query` whenever you can? Or are there situations when this could lead to undesired matches?

I have the feeling we have now lots of options and now rules which one to use. I'm afraid we'll be just experimenting which call makes the tests pass...

If you have some good rules, we should put them in the namespace docstring and/or add an explanation to each of these functions about the correct use. (I know there is already some documentation, but I'm not sure it's (precise) enough.)